File: 501.find-mode-in-binary-search-tree.java
/*
 * @lc app=leetcode id=501 lang=java
 *
 * [501] Find Mode in Binary Search Tree
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int currentCount = 0;
    int maxCount = 0;
    List<Integer> modes = new ArrayList<>();
    int preInorderNodeVal = -1;

    public int[] findMode(TreeNode root) {
        inorder(root);
        int[] res = new int[this.modes.size()];
        for (int i = 0; i < res.length; i++) {
            res[i] = this.modes.get(i);
        }

        return res;
    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }

        this.inorder(root.left);
        
        if (this.preInorderNodeVal == root.val) {
            this.currentCount++;
        } else {
            this.currentCount = 1;
        }

        if (this.currentCount == this.maxCount) {
            this.modes.add(root.val);
        } else if (this.currentCount > this.maxCount) {
            this.modes.clear();
            this.modes.add(root.val);
            this.maxCount = this.currentCount;
        }

        this.preInorderNodeVal = root.val;
        this.inorder(root.right);
    }
}
// @lc code=end

File: 501.find-mode-in-binary-search-tree.py
#
# @lc app=leetcode id=501 lang=python3
#
# [501] Find Mode in Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self) -> None:
        self.modes = []  # Elements with the highest occurrence
        self.current_count = 0
        self.max_count = 0
        self.prev_val = None

    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.inorder(root)
        return self.modes
    
    def inorder(self, root):
        if not root:
            return

        self.inorder(root.left)

        if self.prev_val == root.val:
            self.current_count += 1
        else:
            self.current_count = 1

        if self.current_count == self.max_count:
            self.modes.append(root.val)
        elif self.current_count > self.max_count:
            self.modes = [root.val]
            self.max_count = self.current_count

        self.prev_val = root.val

        self.inorder(root.right)

        
# @lc code=end

